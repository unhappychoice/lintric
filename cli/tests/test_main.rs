use insta::assert_snapshot;
use std::fs;
use std::path::PathBuf;
use std::process::Command;

#[test]
fn test_basic_analysis() {
    let temp_dir = PathBuf::from("tmp");
    fs::create_dir_all(&temp_dir).expect("Unable to create test directory");

    let temp_file_path = temp_dir.join("temp_test_file.rs");
    fs::write(&temp_file_path, "let a = 1;\nlet b = a + 1;\n").expect("Unable to write test file");

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("--verbose")
        .arg(temp_file_path.to_string_lossy().replace("\\", "/"))
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_multiple_files_analysis() {
    let temp_dir = PathBuf::from("tmp");
    fs::create_dir_all(&temp_dir).expect("Unable to create test directory");

    let file1_path = temp_dir.join("file1.rs");
    fs::write(&file1_path, "let x = 1;\nlet y = x + 1;\n").expect("Unable to write file1.rs");

    let file2_path = temp_dir.join("file2.ts");
    fs::write(&file2_path, "const a = 1;\nlet b = a + 1;\n").expect("Unable to write file2.ts");

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg(temp_dir.to_string_lossy().replace("\\", "/"))
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_json_output() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("--json")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_complex_rust_analysis() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("--verbose")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_complex_typescript_analysis() {
    let fixture_path = "../core/tests/typescript/fixtures/complex_typescript_code.ts";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("--verbose")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_ast_rust() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("ast")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_ast_typescript() {
    let fixture_path = "../core/tests/typescript/fixtures/complex_typescript_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("ast")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_definition_rust() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("definition")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_definition_typescript() {
    let fixture_path = "../core/tests/typescript/fixtures/complex_typescript_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("definition")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_dependency_rust() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("dependency")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_dependency_typescript() {
    let fixture_path = "../core/tests/typescript/fixtures/complex_typescript_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("dependency")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}
