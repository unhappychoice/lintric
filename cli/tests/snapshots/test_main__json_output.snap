---
source: cli/tests/test_main.rs
expression: stdout
---
{
  "results": [
    {
      "file_path": "../core/tests/rust/fixtures/complex_rust_code.rs",
      "original_file_path": "../core/tests/rust/fixtures/complex_rust_code.rs",
      "line_metrics": [
        {
          "line_number": 7,
          "total_dependencies": 2,
          "dependency_distance_cost": 0.0625,
          "depth": 1,
          "transitive_dependencies": 1,
          "dependent_lines": [
            6,
            6
          ]
        },
        {
          "line_number": 8,
          "total_dependencies": 1,
          "dependency_distance_cost": 0.03125,
          "depth": 2,
          "transitive_dependencies": 2,
          "dependent_lines": [
            7
          ]
        },
        {
          "line_number": 16,
          "total_dependencies": 6,
          "dependency_distance_cost": 1.46875,
          "depth": 1,
          "transitive_dependencies": 4,
          "dependent_lines": [
            13,
            26,
            12,
            26,
            6,
            6
          ]
        },
        {
          "line_number": 17,
          "total_dependencies": 6,
          "dependency_distance_cost": 1.59375,
          "depth": 2,
          "transitive_dependencies": 5,
          "dependent_lines": [
            13,
            27,
            12,
            27,
            6,
            6
          ]
        },
        {
          "line_number": 22,
          "total_dependencies": 1,
          "dependency_distance_cost": 0.03125,
          "depth": 1,
          "transitive_dependencies": 1,
          "dependent_lines": [
            21
          ]
        },
        {
          "line_number": 23,
          "total_dependencies": 1,
          "dependency_distance_cost": 0.03125,
          "depth": 2,
          "transitive_dependencies": 2,
          "dependent_lines": [
            22
          ]
        },
        {
          "line_number": 27,
          "total_dependencies": 1,
          "dependency_distance_cost": 0.03125,
          "depth": 1,
          "transitive_dependencies": 1,
          "dependent_lines": [
            26
          ]
        },
        {
          "line_number": 28,
          "total_dependencies": 2,
          "dependency_distance_cost": 0.09375,
          "depth": 2,
          "transitive_dependencies": 2,
          "dependent_lines": [
            26,
            27
          ]
        },
        {
          "line_number": 30,
          "total_dependencies": 1,
          "dependency_distance_cost": 0.46875,
          "depth": 1,
          "transitive_dependencies": 1,
          "dependent_lines": [
            15
          ]
        },
        {
          "line_number": 31,
          "total_dependencies": 1,
          "dependency_distance_cost": 0.34375,
          "depth": 1,
          "transitive_dependencies": 1,
          "dependent_lines": [
            20
          ]
        }
      ],
      "overall_complexity_score": 40.415625000000006
    }
  ],
  "total_files_analyzed": 1,
  "total_overall_complexity_score": 40.415625000000006,
  "average_complexity_score": 40.415625000000006
}
